wp-includes/sqlite-ast/class-wp-sqlite-driver.php

```diff
@@ -0,0 +1,1777 @@
+				);
+				error_log( $message );
+			}
+		}
+	}
+
+	/**
+	 * Get the PDO object.
+	 *
+	 * @return PDO
+	 */
+	public function get_pdo() {
+		return $this->pdo;
+	}
+
+	/**
+	 * Method to return inserted row id.
+	 */
+	public function get_insert_id() {
+		return $this->last_insert_id;
+	}
+
+	/**
+	 * Method to return the number of rows affected.
+	 */
+	public function get_affected_rows() {
+		return $this->affected_rows;
+	}
+
+	/**
+	 * Method to execute query().
+	 *
+	 * Divide the query types into seven different ones. That is to say:
+	 *
+	 * 1. SELECT SQL_CALC_FOUND_ROWS
+	 * 2. INSERT
+	 * 3. CREATE TABLE(INDEX)
+	 * 4. ALTER TABLE
+	 * 5. SHOW VARIABLES
+	 * 6. DROP INDEX
+	 * 7. THE OTHERS
+	 *
+	 * #1 is just a tricky play. See the private function handle_sql_count() in query.class.php.
+	 * From #2 through #5 call different functions respectively.
+	 * #6 call the ALTER TABLE query.
+	 * #7 is a normal process: sequentially call prepare_query() and execute_query().
+	 *
+	 * #1 process has been changed since version 1.5.1.
+	 *
+	 * @param string $statement          Full SQL statement string.
+	 * @param int    $mode               Not used.
+	 * @param array  ...$fetch_mode_args Not used.
+	 *
+	 * @see PDO::query()
+	 *
+	 * @throws Exception    If the query could not run.
+	 * @throws PDOException If the translated query could not run.
+	 *
+	 * @return mixed according to the query type
+	 */
+	public function query( $statement, $mode = PDO::FETCH_OBJ, ...$fetch_mode_args ) { // phpcs:ignore WordPress.DB.RestrictedClasses
+		$this->flush();
+		if ( function_exists( 'apply_filters' ) ) {
+			/**
+			 * Filters queries before they are translated and run.
+			 *
+			 * Return a non-null value to cause query() to return early with that result.
+			 * Use this filter to intercept queries that don't work correctly in SQLite.
+			 *
+			 * From within the filter you can do
+			 *  function filter_sql ($result, $translator, $statement, $mode, $fetch_mode_args) {
+			 *     if ( intercepting this query  ) {
+			 *       return $translator->execute_sqlite_query( $statement );
+			 *     }
+			 *     return $result;
+			 *   }
+			 *
+			 * @param null|array $result Default null to continue with the query.
+			 * @param object     $translator The translator object. You can call $translator->execute_sqlite_query().
+			 * @param string     $statement The statement passed.
+			 * @param int        $mode Fetch mode: PDO::FETCH_OBJ, PDO::FETCH_CLASS, etc.
+			 * @param array      $fetch_mode_args Variable arguments passed to query.
+			 *
+			 * @returns null|array Null to proceed, or an array containing a resultset.
+			 * @since 2.1.0
+			 */
+			$pre = apply_filters( 'pre_query_sqlite_db', null, $this, $statement, $mode, $fetch_mode_args );
+			if ( null !== $pre ) {
+				return $pre;
+			}
+		}
+		$this->pdo_fetch_mode = $mode;
+		$this->mysql_query    = $statement;
+		if (
+			preg_match( '/^\s*START TRANSACTION/i', $statement )
```

Comment:
Let's get rid of all the `preg_match()` calls now that we're dealing with ASTs

Reply:
Just marking that this is copy-pasted from the legacy translator. We should definitely fix that 💯